File: observalog-node/src/__tests__/wire.test.ts
import { encodePartA, encodePartB, LogEntry } from '../wire';
import { Err, Outcome } from '../fields';
import * as C from '../wire-contract';

function minimalEntry(): LogEntry {
    return {
        traceId:      'trc_7f2a1b9e4d',
        spanId:       'spn_004',
        parentSpan:   'spn_001',
        seq:          4,
        serviceCode:  C.SERVICE_DOC,
        levelCode:    1,
        outcomeCode:  1,
        tsMs:         1748268153812,
        event:        'doc.storage.saved',
        message:      'Document written to storage',
        durationMs:   undefined,
        outcome:      undefined,
        error:        undefined,
        userId:       '',
        journeyStage: '',
        ctx:          {},
    };
}

describe('Part A encoding', () => {
    it('starts with A:1|', () => {
        expect(encodePartA(minimalEntry())).toMatch(/^A:1\|/);
    });

    it('is exactly 55 content chars + A: prefix + newline = 58 total', () => {
        const line = encodePartA(minimalEntry());
        expect(line.length).toBe(58); // "A:" + 55 content + "\n"
    });

    it('places trace_id at content offset 2 (after A:)', () => {
        const line = encodePartA(minimalEntry());
        // Offsets are within the content after the "A:" prefix.
        const payload = line.slice(2); // skip "A:"
        const traceId = payload.slice(C.TRACE_ID_OFFSET, C.TRACE_ID_OFFSET + 14).trimEnd();
        expect(traceId).toBe('trc_7f2a1b9e4d');
    });

    it('encodes seq as 2-char lowercase hex', () => {
        const e = { ...minimalEntry(), seq: 255 };
        const line = encodePartA(e);
        const payload = line.slice(2); // skip "A:"
        const seqField = payload.slice(C.SEQ_OFFSET, C.SEQ_OFFSET + 2);
        expect(seqField).toBe('ff');
    });

    it('uses "-------" for absent parent_span', () => {
        const e = { ...minimalEntry(), parentSpan: '' };
        const line = encodePartA(e);
        const payload = line.slice(2); // skip "A:"
        const parent = payload.slice(C.PARENT_SPAN_OFFSET, C.PARENT_SPAN_OFFSET + 7);
        expect(parent).toBe('-------');
    });
});

describe('Part B encoding', () => {
    it('abbreviates event to "e"', () => {
        const line = encodePartB(minimalEntry());
        const parsed = JSON.parse(line.trim());
        expect(parsed).toHaveProperty('e');
        expect(parsed).not.toHaveProperty('event');
    });

    it('abbreviates message to "m"', () => {
        const parsed = JSON.parse(encodePartB(minimalEntry()).trim());
        expect(parsed).toHaveProperty('m', 'Document written to storage');
    });

    it('encodes error struct with abbreviated keys', () => {
        const e = {
            ...minimalEntry(),
            error: new Err('RateLimitExceeded', 'QUOTA', 'limit hit', true),
        };
        const parsed = JSON.parse(encodePartB(e).trim());
        expect(parsed.er).toMatchObject({ ek: 'RateLimitExceeded', ec: 'QUOTA', rt: true });
    });

    it('encodes outcome string and duration_ms', () => {
        const e = { ...minimalEntry(), outcome: Outcome.Failure, durationMs: 87 };
        const parsed = JSON.parse(encodePartB(e).trim());
        expect(parsed.o).toBe('failure');
        expect(parsed.ms).toBe(87);
    });
});

File: observalog-java/src/test/java/com/observalog/WireTest.java
package com.observalog;

import org.junit.jupiter.api.Test;
import java.io.IOException;
import java.util.Collections;

import static org.junit.jupiter.api.Assertions.*;

class WireTest {

    private LogEntry minimalEntry() {
        LogEntry e = new LogEntry();
        e.traceId      = "trc_7f2a1b9e4d";
        e.spanId       = "spn_004";
        e.parentSpan   = "spn_001";
        e.seq          = 4;
        e.serviceCode  = WireContract.SERVICE_DOC;
        e.levelCode    = 1; // INFO
        e.outcomeCode  = 1; // success
        e.tsMs         = 1748268153812L;
        e.event        = "doc.storage.saved";
        e.message      = "Document written to storage";
        e.userId       = "";
        e.journeyStage = "";
        e.ctx          = Collections.emptyMap();
        return e;
    }

    @Test
    void partA_prefix() {
        byte[] partA = Wire.encodePartA(minimalEntry());
        String line = new String(partA);
        assertTrue(line.startsWith("A:1|"), "Part A must start with A:1|");
    }

    @Test
    void partA_traceId_at_offset_2() {
        byte[] partA = Wire.encodePartA(minimalEntry());
        // Skip "A:" prefix (2 bytes)
        String payload = new String(partA).substring(2);
        String traceId = payload.substring(
            WireContract.TRACE_ID_OFFSET - 2,
            WireContract.SPAN_ID_OFFSET - 2 - 1  // -1 for separator
        ).stripTrailing();
        assertEquals("trc_7f2a1b9e4d", traceId);
    }

    @Test
    void partA_seq_hex_encoded() {
        LogEntry e = minimalEntry();
        e.seq = 255;
        String line = new String(Wire.encodePartA(e));
        // seq field is at positions 33-34 after the "A:" prefix
        String seqField = line.substring(2 + WireContract.SEQ_OFFSET, 2 + WireContract.SEQ_OFFSET + 2);
        assertEquals("ff", seqField);
    }

    @Test
    void partA_absent_parent_span_uses_dashes() {
        LogEntry e = minimalEntry();
        e.parentSpan = "";
        String line = new String(Wire.encodePartA(e));
        // parent span at offsets 25-31 after "A:" prefix
        String parentField = line.substring(2 + WireContract.PARENT_SPAN_OFFSET,
                                             2 + WireContract.PARENT_SPAN_OFFSET + 7);
        assertEquals("-------", parentField);
    }

    @Test
    void partA_length_is_55_plus_newline() {
        byte[] partA = Wire.encodePartA(minimalEntry());
        // Content is 55 chars + "A:" prefix (2) + newline (1) = 58
        assertEquals(58, partA.length);
    }

    @Test
    void partB_has_abbreviated_event_key() throws IOException {
        LogEntry e = minimalEntry();
        String partB = new String(Wire.encodePartB(e));
        assertTrue(partB.contains("\"e\":"), "event must be abbreviated to 'e'");
        assertFalse(partB.contains("\"event\":"), "full key 'event' must not appear");
    }

    @Test
    void partB_error_struct_uses_abbreviated_keys() throws IOException {
        LogEntry e = minimalEntry();
        e.error = new Err("RateLimitExceeded", "QUOTA", "limit hit", true);
        String partB = new String(Wire.encodePartB(e));
        assertTrue(partB.contains("\"ek\""), "error.kind → ek");
        assertTrue(partB.contains("\"ec\""), "error.code → ec");
        assertTrue(partB.contains("\"rt\""), "error.retryable → rt");
    }
}
